src/mem/ruby/protocol/chi/tlm/controller.cc

```diff
@@ -0,0 +1,424 @@
+        recvDBID = true;
+    }
+
+    phase.dbid = msg->m_dbid % 1024;
+    Transaction::handle(msg);
+
+    return recvComp && recvDBID;
+}
+
+void
+CacheController::sendCompAck(ARM::CHI::Payload &payload,
+                             ARM::CHI::Phase &phase)
+{
+    auto res_msg = std::make_shared<CHIResponseMsg>(
+        curTick(), cacheLineSize, m_ruby_system);
+
+    res_msg->m_addr = ruby::makeLineAddress(payload.address, cacheLineBits);
+    res_msg->m_type = CHIResponseType_CompAck;
+    res_msg->m_Destination.add(mapAddressToDownstreamMachine(payload.address));
+    res_msg->m_responder = getMachineID();
+
+    res_msg->m_txnId = phase.txn_id;
+
+    sendResponseMsg(res_msg);
+}
+
+void
+CacheController::sendMsg(ARM::CHI::Payload &payload, ARM::CHI::Phase &phase)
+{
+    switch (phase.channel) {
+      case ARM::CHI::CHANNEL_REQ:
+        sendRequestMsg(payload, phase);
+        break;
+      case ARM::CHI::CHANNEL_DAT:
+        sendDataMsg(payload, phase);
+        break;
+      case ARM::CHI::CHANNEL_RSP:
+        sendResponseMsg(payload, phase);
+        break;
+      default:
+        panic("Unimplemented channel: %d", phase.channel);
+    }
+}
+
+Addr
+CacheController::reqAddr(ARM::CHI::Payload &payload,
+                         ARM::CHI::Phase &phase) const
+{
+    switch (phase.req_opcode) {
+      case ARM::CHI::REQ_OPCODE_WRITE_NO_SNP_PTL:
+        return ruby::makeLineAddress(payload.address, cacheLineBits) +
+            ctz64(payload.byte_enable);
+      default:
+        return payload.address;
+    }
+}
+
+Addr
+CacheController::reqSize(ARM::CHI::Payload &payload,
+                         ARM::CHI::Phase &phase) const
+{
+    switch (phase.req_opcode) {
+      case ARM::CHI::REQ_OPCODE_WRITE_NO_SNP_PTL:
+        assert(transactionSize(payload.size) >= popCount(payload.byte_enable));
+        return popCount(payload.byte_enable);
+      default:
+        return transactionSize(payload.size);
+    }
+}
+
+void
+CacheController::sendRequestMsg(ARM::CHI::Payload &payload,
+                                ARM::CHI::Phase &phase)
+{
+    auto req_msg = std::make_shared<CHIRequestMsg>(
+        curTick(), cacheLineSize, m_ruby_system);
+
+    req_msg->m_addr = ruby::makeLineAddress(payload.address, cacheLineBits);
+    req_msg->m_accAddr = reqAddr(payload, phase);
+    req_msg->m_accSize = reqSize(payload, phase);
+    req_msg->m_requestor = getMachineID();
+    req_msg->m_fwdRequestor = getMachineID();
+    req_msg->m_dataToFwdRequestor = false;
+    req_msg->m_type = tlm_to_ruby::reqOpcode(phase.req_opcode);
+    req_msg->m_isSeqReqValid = false;
+    req_msg->m_is_local_pf = false;
+    req_msg->m_is_remote_pf = false;
+    req_msg->m_allowRetry = phase.allow_retry;
+    req_msg->m_Destination.add(mapAddressToDownstreamMachine(payload.address));
+
+    req_msg->m_txnId = phase.txn_id + (payload.lpid * 1024);
+    req_msg->m_ns = payload.ns;
+
+    sendRequestMsg(req_msg);
```

Comment:
This (and the other `send*Msg` functions) currently assumes the underlying MessageBuffer is configured with unlimited capacity. I don't think this is a blocker for this PR, but it would be good to have a followup PR that adds some sort of link-level credit management to properly handle backpressure.

For instance, in CHIGenericController we could have a `virtual bool recvLinkCrd(unsigned crds, CHIChannel channel)` (or a separate one per channel type) that top-level can optionally implement to receive credits. It could be called from `CHIGenericController::init or initNetQueues` to return the initial number of credits (or 0 or UINT_MAX if the buffer is infinite), and every time a message is dequeued. We would need to register a callback with `MessageBuffer::registerDequeueCallback` to be informed of a dequeue.

I can create an issue to track this once it is merged.